packages/tailwind/src/tailwind.tsx

```diff
@@ -12,133 +13,91 @@ export interface TailwindProps {
+
+  const cloneElementWithInlinedStyles =
+    useCloneElementWithInlinedStyles(tailwind);
+
+  const nonInlineStylesRootToApply = new Root();
+  let mediaQueryClassesForAllElement: string[] = [];
+  let hasNonInlineStylesToApply = false as boolean;
+
+  let mappedChildren = mapReactTree(children, (node) => {
+    if (React.isValidElement<EmailElementProps>(node)) {
+      const element = node;
+
+      const {
+        elementWithInlinedStyles,
+        nonInlinableClasses,
+        nonInlineStyleNodes,
+      } = cloneElementWithInlinedStyles(element);
+      mediaQueryClassesForAllElement =
+        mediaQueryClassesForAllElement.concat(nonInlinableClasses);
+      nonInlineStylesRootToApply.append(nonInlineStyleNodes);
+
+      if (nonInlinableClasses.length > 0 && !hasNonInlineStylesToApply) {
+        hasNonInlineStylesToApply = true;
       }
-    }
-
-    if (element.props.children) {
-      propsToOverwrite.children = React.Children.map(
-        element.props.children,
-        (child) => {
-          if (React.isValidElement<EmailElementProps>(child)) {
-            return processElement(child);
-          }
 
-          return child;
-        },
-      );
+      return elementWithInlinedStyles;
     }
 
-    if (element.props.className) {
-      const { styles, residualClassName } = inline(element.props.className);
-      propsToOverwrite.style = {
-        ...element.props.style,
-        ...styles,
-      };
-      if (residualClassName.trim().length > 0) {
-        propsToOverwrite.className = residualClassName;
-        /*
-          We sanitize only the class names of Tailwind classes that we are not going to inline
-          to avoid unpredictable behavior on the user's code. If we did sanitize all class names
-          a user-defined class could end up also being sanitized which would lead to unexpected 
-          behavior and bugs that are hard to track.
-        */
-        for (const singleClass of nonInlinableClasses) {
-          propsToOverwrite.className = propsToOverwrite.className.replace(
-            singleClass,
-            sanitizeClassName(singleClass),
-          );
-        }
-      } else {
-        propsToOverwrite.className = undefined;
-      }
-    }
+    return node;
+  });
 
-    const newProps = {
-      ...element.props,
-      ...propsToOverwrite,
-    };
-    const newChildren = propsToOverwrite.children
-      ? propsToOverwrite.children
-      : element.props.children;
-
-    if (typeof element.type === "function") {
-      const component = element.type as React.FC;
-      const renderedComponent = component({
-        ...element.props,
-        ...propsToOverwrite,
-      });
-
-      if (React.isValidElement<EmailElementProps>(renderedComponent)) {
-        return processElement(renderedComponent);
+  if (hasNonInlineStylesToApply) {
+    let hasAppliedNonInlineStyles = false as boolean;
+    mappedChildren = mapReactTree(mappedChildren, (node) => {
```

Comment:
Following the same idea from what we merged recently that allows the `<head>` element to be at any depth, this maps the entire React tree to find it.